### ISSUE TITLE
Use cases for backend interoperability

### ISSUE THREAD

**Issue by AerialMantis** [KhronosGroup/SYCL-Docs#230]
SYCL 2020 provides an optional interface for application scope interoperability with the native backend API, where different implementations can choose which classes to provide interoperability for. Backend specifications can also then define which classes should have interoperability for an implementation of that backend.

This leads to the question of which classes should to provide interoperability for, and the best way to answer this question is to derive it from the possible use cases of application scope backend interoperability.

The current use cases we have identified are:

* Interoperability at the SYCL DAG level, using host tasks to inject arbitrary C++ functions with backend specific API calls between SYCL commands, a typical example of this would be the oneDNN project (https://github.com/oneapi-src/oneDNN).
* Interoperability at the kernel level, using a kernel_bundle to create a SYCL kernel from a backend-specific source code or IR and running that via the SYCL runtime.
* Interoperability with graphics APIs, sharing memory objects between SYCL the native backend API in order to interoperate with a graphics API such as DirectX or Vulkan for post/pre-processing.

Does anyone know of other use cases for SYCL application scope backend interoperability that we've not considered here?

**Comment by illuhad** [comment]
@tomdeakin Are you asking for our https://github.com/illuhad/hipSYCL/blob/develop/doc/enqueue-custom-operation.md extension? ;) I agree that host tasks are not the right way for e.g. cuBLAS.

I believe `sycl::queue::get_native()` is not such a good idea for this purpose, since it does not work in cases where there's not a clear mapping between backend queues or SYCL queues (i.e. hipSYCL, which uses a queue pool with global scheduler) or when the backend is not queue-based at all. The current backend interop API does not require implementations to be available to provide backend interop for every SYCL type.
An approach as in the hipSYCL extension allows to append external device tasks to the task graph (instead of host tasks), and it works regardless of how the SYCL implementation maps queues to backend objects.

**Comment by tomdeakin** [comment]
I've been spending a **lot** of time reading up on OpenMP's approach to interoperating. They will have a similar problem to what you describe in that the implementations use a queue pool. Their solution is to provide asynchronous synchronisation guarantees to ensure a "happens before" ordering of kernels.

For SYCL with a CUDA backend, the implementation could insert `cudaStreamWaitEvent()` on all queues in the pool with some event created when you get a native queue as part of the `sycl::queue::get_native()` implementation. You'd need to do the same via some mechanism to *return* the native queue back to the pool. This would mean that all SYCL stuff would finish, the work on the native queue would run, and then SYCL stuff could carry on again. Meanwhile, the host has been busy submitting (to wherever) and is not stuck in a synchronous call.

What we might be missing is a `SYCL::queue::return_native()` API. This would give you the semantics to help define an ordering.

OpenMP doesn't define an ordering of enqueued OpenMP tasks and external tasks. For example where you submit both to the regular DAG and the external native queue simultaneously, there is no way to synchronise. But you do get some API for a "line in the sand" where you can reason about ordering.

**Comment by illuhad** [comment]
@tomdeakin I don't get it. Our extension allows you do just that, getting a native queue-like type (or whatever the backend uses) to which users can enqueue without having to wait on things to complete, and having any submitted operations synchronize with the DAG automagically.

I don't understand how `sycl::queue::return_native()` would be different from `get_native<queue>()` that we already have. Would this approach help you if you're not on a backend that relies on in-order queues?

> For SYCL with a CUDA backend, the implementation could insert cudaStreamWaitEvent() on all queues in the pool

Wouldn't synchronizing all queues be super inefficient?

I'm not sure I follow what you want to achieve and how this is different from e.g. what hipSYCL can already do with its extension. Can you maybe provide an example?

**Comment by tomdeakin** [comment]
@illuhad - By `return_native()` I meant "give back" the queue obtained the `get_native`. So you have a pair which define some ordering.

I would hope that asynchronously synchronising streams with `cudaStreamWaitEvent` was cheap, especially if by the time that wait event runs, all work has already finished.

The HipSYCL extension looks good. I think it would solve most of the requirements. Your extension provides exactly ordering I described: you are guaranteed that: SYCL stuff that was enqueued before you called `enqueue_custom_operation()` will run before the native calls enqueued inside that function, and SYCL stuff enqueued after you called `enqueue_custom_operation()` will run after the native calls enqueued inside that function.

**Comment by illuhad** [comment]
@tomdeakin Okay, I think I understand now. I think one big problem for interop is that we *still* need a mechanism to get native memory objects (e.g. device pointers) from buffers in order to be able to do useful interop. Due to lazy allocation of buffers in SYCL, this is only possible by going through the SYCL runtime with an operation passed to `submit()`. If we add this requirement, you end up pretty much at the hipSYCL extension :)
Additionally, `cudaStreamWaitEvent` might not be so cheap when a global stream pool is involved as in hipSYCL, since then for any interop operation the execution of all SYCL might be blocked. Consequently, I think that such an interop mechanism must live as part of the DAG (as in the hipSYCL extension).

**Comment by tomdeakin** [comment]
Ok, the hipSYCL extension looks like it’s already thought of that too!

**Comment by illuhad** [comment]
@npmiller I don't think it necessarily does because you can absorb additional behavior into the actual interop type that is exposed in the `custom_operation`. But you are right that e.g. for OpenCL you might additionally want a mechanism to feed an event back into the SYCL runtime.

**Comment by illuhad** [comment]
> So in a more general sense, does the get_native interface for sycl::buffer provided in the SYCL 2020 spec really makes sense then? Maybe some vendors and/or backends are still able to provide it.

Good question. Maybe if the buffer is only bound to a single context, and you now that your implementation might just store a `cl_mem` or something like that... In general it would probably require in-depth knowledge about implementation behavior to be useful.

IMO we are also lacking a way to force preallocation on a device to make this more predictable, see e.g. https://github.com/illuhad/hipSYCL/blob/develop/doc/extensions.md#hipsycl_ext_update_device

**Comment by keryell** [comment]
I cannot see in the list above how to call a native function from a SYCL kernel.
